File: examples/javascript/quickstart.mjs
/**
 * Copyright 2026 Volstrata.com
 *
 * The smallest useful call: one Free capability, no key, no install, no setup.
 * Docs: https://volstrata.com/docs/api-examples
 *
 * Run:  node quickstart.mjs [TICKER]
 *       npm run quickstart
 *
 * Requires Node 18 or newer — that is where global `fetch` arrives. Nothing is
 * installed: this folder has no runtime dependencies, and `gex.levels` answers
 * an anonymous caller, so there is nothing to set up before running it.
 *
 * One request is made.
 */

import { BASE, VolstrataApiError, apiFetch, hasApiKey, heading, show, showRateLimit } from "./lib/volstrata.mjs";

const ticker = (process.argv[2] ?? "SPX").toUpperCase();

heading("VolStrata quickstart");
show("base url", BASE);
show("capability", `gex.levels  →  GET /api/v1/gex/levels?ticker=${ticker}`);
show("api key", hasApiKey ? "detected in VOLSTRATA_API_KEY" : "none — this call does not need one");

try {
  // `apiFetch` returns both the Response and the decoded body, because the
  // headers are worth reading too — see the rate-limit line printed below.
  //
  // The HTTP status is checked before the body is touched. That ordering is the
  // one habit to copy from this file: a refusal has a completely different body
  // shape, so branching on a field like `ok` would read a field that is not
  // there. Status first, always.
  const { res, data } = await apiFetch("/api/v1/gex/levels", { ticker });

  heading("Response");
  show("ok", data.ok); // true on success — a confirmation, not the success test
  show("ticker", data.ticker);
  show("spot", data.spot); // last price the levels were computed against
  show("updated", data.updated); // human-readable, exchange-local
  show("ts", data.ts); // the same instant as a unix timestamp

  // `levels` is a flat map of named price levels. Keys are stable, but any one
  // of them can be null when that level is not defined for the current session,
  // so read them defensively instead of indexing straight in. What each name
  // means is in the glossary: https://volstrata.com/docs/api-catalog
  const levels = data.levels ?? {};
  heading("A few named levels");
  for (const key of ["cw", "pw", "zg", "mp", "emHigh", "emLow", "hvl"]) {
    if (key in levels) show(key, levels[key] ?? "null (not defined right now)");
  }
  show("keys returned", Object.keys(levels).length);

  heading("Budget");
  showRateLimit(res); // every response carries these, success or failure

  heading("Next");
  console.log("  node auth_and_errors.mjs   keys, RFC 9457 errors, plan floors");
  console.log("  node pagination.mjs        cursor paging over the capability catalog");
  console.log("  node rest/gex_levels.mjs   the gex.* domain, end to end");
  if (!hasApiKey) {
    console.log("\n  Everything above ran anonymously. A key unlocks the capabilities");
    console.log("  above the Free floor: https://volstrata.com/api-keys");
  }
} catch (error) {
  // A Free capability failing usually means the network, not the API. Print what
  // is actually known rather than guessing.
  heading("The call did not succeed");
  if (error instanceof VolstrataApiError) {
    console.log(error.describe("  "));
    if (!error.isApiResponse) {
      console.log("\n  A non-JSON refusal never reached the API. Check the User-Agent");
      console.log("  header and any proxy between you and volstrata.com.");
    }
  } else {
    console.log(`  ${error?.message ?? error}`);
  }
  console.log("\n  Errors reference: https://volstrata.com/docs/api-errors");
  process.exitCode = 1;
}

File: examples/python/quickstart.py
"""Copyright 2026 Volstrata.com

The shortest thing that works: one unauthenticated call to the VolStrata API.
Docs: https://volstrata.com/docs/api-examples

    python -m pip install requests
    python quickstart.py

No API key. No .env file. No account. `gex.levels` answers an anonymous caller,
so this runs the moment you have `requests` installed.

When you do have a key, export it and every example in this folder picks it up:

    export VOLSTRATA_API_KEY=...        # macOS / Linux
    $env:VOLSTRATA_API_KEY = "..."      # Windows PowerShell
"""

import sys

from volstrata_helpers import (
    API,
    VolstrataError,
    get,
    have_key,
    heading,
    row,
)


def main() -> int:
    print("VolStrata quickstart -- https://volstrata.com")
    print("Base: {0}".format(API))

    # GET /api/v1/gex/levels?ticker=SPX
    #
    # `ticker` defaults to SPX, so passing it is optional here. It is passed
    # explicitly because that is the habit you want in your own code.
    try:
        data = get("/gex/levels", ticker="SPX")
    except VolstrataError as exc:
        # Anything 4xx/5xx arrives as an RFC 9457 problem document, already
        # parsed. auth_and_errors.py walks through each kind.
        print("\nThe call was refused: {0}".format(exc))
        return 1

    # A successful response carries `ok: true` alongside the payload. Do not use
    # it as your success test: the HTTP status is the source of truth, and
    # volstrata_helpers.get() has already checked it before returning.
    heading("gex.levels")
    row("ok", data.get("ok"))
    row("ticker", data.get("ticker"))
    row("spot", data.get("spot"))
    row("updated", data.get("updated"))       # human-readable, exchange-local
    row("ts", data.get("ts"))                 # the same instant, unix seconds

    # `levels` is a flat map of named price levels. Every key is defined in the
    # public glossary; GET /api/v1/docs/metrics lists them, and
    # https://volstrata.com/docs/api-catalog documents the capability itself.
    # A key can be null when that level is not defined for the current session,
    # so read defensively rather than indexing straight in.
    levels = data.get("levels") or {}
    heading("A few named levels")
    for key in ("cw", "pw", "zg", "mp", "emHigh", "emLow"):
        if key in levels:
            row(key, levels[key])
    row("keys returned", len(levels))

    heading("Your setup")
    if have_key():
        print("  VOLSTRATA_API_KEY is set -- the call above was authenticated.")
        print("  Try auth_and_errors.py next to see what a key changes.")
    else:
        print("  No VOLSTRATA_API_KEY set. The call above ran anonymously.")
        print("  Issue a key at https://volstrata.com/api-keys when you want the")
        print("  capabilities behind a plan floor.")

    print("\nNext: python discover_capabilities.py   (what else is published)")
    print("      python auth_and_errors.py         (keys, errors, rate limits)")
    return 0


if __name__ == "__main__":
    sys.exit(main())
